src/webgpu/shader/execution/expression/binary/af_subtraction.spec.ts

```diff
@@ -0,0 +1,151 @@
+export const description = `
+Execution Tests for non-matrix AbstractFloat subtraction expression
+`;
+
+import { makeTestGroup } from '../../../../../common/framework/test_group.js';
+import { GPUTest } from '../../../../gpu_test.js';
+import { TypeAbstractFloat, TypeVec } from '../../../../util/conversion.js';
+import { FP, FPVector } from '../../../../util/floating_point.js';
+import { sparseF64Range, sparseVectorF64Range } from '../../../../util/math.js';
+import { makeCaseCache } from '../case_cache.js';
+import { onlyConstInputSource, run } from '../expression.js';
+
+import { abstractBinary } from './binary.js';
+
+const subtractionVectorScalarInterval = (v: number[], s: number): FPVector => {
+  return FP.abstract.toVector(v.map(e => FP.abstract.subtractionInterval(e, s)));
+};
+
+const subtractionScalarVectorInterval = (s: number, v: number[]): FPVector => {
+  return FP.abstract.toVector(v.map(e => FP.abstract.subtractionInterval(s, e)));
+};
+
+export const g = makeTestGroup(GPUTest);
+
+export const d = makeCaseCache('abstractBinary/af_subtraction', {
+  scalar: () => {
+    return FP.abstract.generateScalarPairToIntervalCases(
+      sparseF64Range(),
+      sparseF64Range(),
+      'finite',
+      FP.abstract.subtractionInterval
+    );
+  },
+  vec2_scalar: () => {
+    return FP.abstract.generateVectorScalarToVectorCases(
+      sparseVectorF64Range(2),
+      sparseF64Range(),
+      'finite',
+      subtractionVectorScalarInterval
+    );
+  },
+  vec3_scalar: () => {
+    return FP.abstract.generateVectorScalarToVectorCases(
+      sparseVectorF64Range(3),
+      sparseF64Range(),
+      'finite',
+      subtractionVectorScalarInterval
+    );
+  },
+  vec4_scalar: () => {
+    return FP.abstract.generateVectorScalarToVectorCases(
+      sparseVectorF64Range(4),
+      sparseF64Range(),
+      'finite',
+      subtractionVectorScalarInterval
+    );
+  },
+  scalar_vec2: () => {
+    return FP.abstract.generateScalarVectorToVectorCases(
+      sparseF64Range(),
+      sparseVectorF64Range(2),
+      'finite',
+      subtractionScalarVectorInterval
+    );
+  },
+  scalar_vec3: () => {
+    return FP.abstract.generateScalarVectorToVectorCases(
+      sparseF64Range(),
+      sparseVectorF64Range(3),
+      'finite',
+      subtractionScalarVectorInterval
+    );
+  },
+  scalar_vec4: () => {
+    return FP.abstract.generateScalarVectorToVectorCases(
+      sparseF64Range(),
+      sparseVectorF64Range(4),
+      'finite',
+      subtractionScalarVectorInterval
+    );
+  },
+});
+
+g.test('scalar')
```

Comment:
Actually this test covers not only scalar-scalar cases but also vector-vector cases by vectorization. Not sure if we would need to change this name?

Reply:
Hmmm... naming things is hard :smile:
I agree that the name scalar doesn't convey that scalar-scalar and vector-vector tests are occuring. The name comes from these being named things like 'f16' and 'f32' in other test files.

The sticking point for a new name I am hitting is that fact that we want something that indicates that homogenous operations are occuring, but it not be so generic as to suggest that it might also cover matrix-matrix operations.

I am tempted to separate these up into two test fixtures 'scalar' and 'vector', and only use `.combine('vectorize', [2, 3, 4] as const)` on 'vector'

I think this requires a bit of thinking to decide on the correct change, so I am going to leave the naming as-is for these PRs, and file a new issue (https://github.com/gpuweb/cts/issues/2935), since this will affect the f32 and f16 tests too